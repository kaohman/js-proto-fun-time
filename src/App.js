import React, { Component } from 'react';
import './App.css';
import './styles/normalize.css';
import './styles/main.scss';
import Step1 from './Step1.js';
import Step2 from './Step2.js';
import Step3 from './Step3.js';
import Instructions from './Instructions.js';

class App extends Component {
  constructor() {
    super();
    this.state = {
      step: 3,
      questionCount: 0,
      showInstructions: false,
      problems: [],
      currentProblem: {}
    }
  }

  toggleInstructionsCard = () => {
    this.setState({
      showInstructions: !this.state.showInstructions,
      problems: []
    });
  }

  incrementStep = () => {
    let newStep = this.state.step < 3 ? this.state.step + 1 : 1;
    newStep === 1 && this.updateQuestion();
    this.setState({
      step: newStep
    });
  }

  updateQuestion = () => {
    let newQuestionCount = this.state.questionCount + 1;
    console.log(newQuestionCount)
    this.setState({
      questionCount: newQuestionCount,
      currentProblem: this.state.problems[newQuestionCount]
    });
  }

  componentDidMount() {
    fetch('http://memoize-datasets.herokuapp.com/api/v1/problems')
      .then(data => data.json())
      .then(results => {
        let randomResults = results.problems.sort((a, b) => 0.5 - Math.random());
        this.setState({
          problems: randomResults,
          currentProblem: randomResults[0]
        });
      })
      .catch(error => console.log(error));
  }

  render() {
    let { step, showInstructions, currentProblem, problems, questionCount } = this.state;
    return (
      <div>
        <header>
          <h2>
            Question {questionCount+1} of {problems.length}
            <span id='difficulty-text'>Difficulty: {currentProblem.difficulty}</span>
          </h2>
          <h1>[ jsProtoFunTime ]</h1>
          <button onClick={this.toggleInstructionsCard} id="instructions-button">
          {
            showInstructions ? 'Hide Instructions' : 'Show Instructions'
          }
          </button>
        </header>
        <div className='game-container'>
          <Step1 incrementStep={this.incrementStep} question={currentProblem.question} currentStep={step}/>
          {
            step > 1 && <Step2 incrementStep={this.incrementStep} correctMethod={currentProblem.method} currentStep={step}/>
          }
          {
            step > 2 && 
            <Step3 
              incrementStep={this.incrementStep} 
              correctAnswer={currentProblem.result}
              input={currentProblem.input}
              currentStep={step}
            />
          }
        </div>
        {
          showInstructions && <Instructions />
        }
      </div>
    );
  }
}

export default App;
