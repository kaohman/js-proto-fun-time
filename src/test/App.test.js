import React from 'react';
import ReactDOM from 'react-dom';
import App from '../App';
import { shallow } from 'enzyme';

const problemsMock = [{
  question: "You are given the following array of numbers: [1, 3, 5, 6, 7, 9]. You are only interested in numbers that are divisible by 3. Return a new array or numbers that contains only the numbers divisible by 3.",
  inputDataType: "array",
  outputDataType: "array",
  method: "filter",
  input: [1, 3, 5, 6, 7, 9],
  result: [3, 6, 9],
  difficulty: 2
  },
  {
    question: "You are given the following array of words to use in Scrabble: ['dog', 'fright', 'silly', 'chair', 'kittens', 'pizza']. You only want to use a few of your letters though so you would like to filter out the longer words. Return a new array of only words with fewer than 5 letters.",
    inputDataType: "array",
    outputDataType: "array",
    method: "filter",
    input: ["dog", "fright", "silly", "chair", "kittens", "pizza"],
    result: ["dog", "silly", "chair", "pizza"],
    difficulty: 3
  },
  {
    question: "You are given the following array of numbers: [-1, 8, 3, -5, 2, 9]. You don't only like positive numbers so you want to filter out all the negativity. Return a new array with only positive numbers.",
    inputDataType: "array",
    outputDataType: "array",
    method: "filter",
    input: [-1, 8, 3, -5, 2, 9],
    result: [8, 3, 2, 9],
    difficulty: 2
  }
];

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe('App', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <App />
    );

    wrapper.setState({
      step: 1,
      questionCount: 0,
      showInstructions: false,
      problems: problemsMock,
      totalProblems: 0,
      currentProblem: problemsMock[0],
      loaded: true
    });
  });

  it('should match the snapshot with all data passed in', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a default state', () => {
    expect(wrapper.state()).toEqual({ 
      step: 1,
      questionCount: 0,
      showInstructions: false,
      problems: problemsMock,
      totalProblems: 0,
      currentProblem: problemsMock[0],
      loaded: true
     });
  });

  it('should toggle showInstructions in state when toggleInstructionsCard if called', () => {
    expect(wrapper.state('showInstructions')).toEqual(false);
    wrapper.find('#instructions-button').simulate('click');
    expect(wrapper.state('showInstructions')).toEqual(true);
  });

  it('should increment the step in state when incrementStep is called', () => {
    wrapper.instance().incrementStep();
    expect(wrapper.state('step')).toEqual(2);
    wrapper.instance().incrementStep();
    expect(wrapper.state('step')).toEqual(3);
    wrapper.instance().incrementStep();
    expect(wrapper.state('step')).toEqual(1);
  });

  it('should update the questionCount and currentProblem in state when updateQuestion is called', () => {
    expect(wrapper.state('questionCount')).toEqual(0);
    expect(wrapper.state('currentProblem')).toEqual(problemsMock[0]);
    wrapper.instance().updateQuestion();
    expect(wrapper.state('questionCount')).toEqual(1);
    expect(wrapper.state('currentProblem')).toEqual(problemsMock[1]);
  });

  it('should update the questionCount, problems, and currentProblem in state when updateGame is called', () => {
    wrapper.setState( { step: 3, questionCount: 3 } );
    wrapper.instance().updateGame();
    expect(wrapper.state('questionCount')).toEqual(0);
    expect(wrapper.state('problems').length).toEqual(3);
  });

});