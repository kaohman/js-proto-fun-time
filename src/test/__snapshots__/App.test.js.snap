// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot with all data passed in 1`] = `
<div>
  <header>
    <h2>
      Question 
      1
       of 
      0
      <span
        id="difficulty-text"
      >
        Difficulty: 
        2
      </span>
    </h2>
    <h1>
      [ jsProtoFunTime ]
    </h1>
    <button
      id="instructions-button"
      onClick={[Function]}
    >
      Show Instructions
    </button>
  </header>
  <div
    className="game-container"
  >
    <Step1
      currentStep={1}
      incrementStep={[Function]}
      question="You are given the following array of numbers: [1, 3, 5, 6, 7, 9].

You are only interested in numbers that are divisible by 3. Return a new array or numbers that contains only the numbers divisible by 3."
    />
  </div>
</div>
`;
